File: src/eqn/ast/EqnAstNodePreDefinedUnaryFunction.java
package eqn.ast;

import eqn.ast.predefined.EqnAstNodeCos;
import eqn.ast.predefined.EqnAstNodeExp;
import eqn.ast.predefined.EqnAstNodeSin;
import eqn.ast.predefined.EqnAstNodeTan;
import eqn.parser.exception.EqnException;
import org.jetbrains.annotations.NotNull;

public abstract class EqnAstNodePreDefinedUnaryFunction extends EqnAstNode {

    protected EqnAstNode operand;

    public EqnAstNodePreDefinedUnaryFunction(String value, @NotNull EqnAstNode equationNode) {
        super(value, EqnAstNode.Type.PreDefinedUnaryFunction);
        this.operand = equationNode;
    }

    public static EqnAstNodePreDefinedUnaryFunction create(String value, @NotNull EqnAstNode equationNode) throws EqnException {
        if (value.equals("sin")) {
            return new EqnAstNodeSin(value, equationNode);
        }
        if (value.equals("cos")) {
            return new EqnAstNodeCos(value, equationNode);
        }
        if (value.equals("tan")) {
            return new EqnAstNodeTan(value, equationNode);
        }
        if (value.equals("exp")) {
            return new EqnAstNodeExp(value, equationNode);
        }
        throw new EqnException("PreDefinedUnaryFunction " + value + " not yet implemented!");
    }

    @Override
    public String toString() {
        return value + '(' + operand.toString() + ')';
    }

    @Override
    public EqnAstNode simplify() throws EqnException {
        operand = operand.simplify();
        if (operand.type == Type.Constant) {
            return new EqnAstNodeDouble(this.evaluate());
        }
        return this;
    }
}

File: src/eqn/ast/EqnAstNodePreDefinedUnaryFunction.kt
package eqn.ast

import eqn.parser.exception.EqnException

abstract class EqnAstNodePreDefinedUnaryFunction(value: String, equationNode: EqnAstNode) : EqnAstNodeUnaryFunction(value, Type.PreDefinedUnaryFunction, equationNode, PrecedenceType.Function) {
    override fun toString(): String {
        return value + '(' + operand().toString() + ')'
    }

    @Throws(EqnException::class)
    override fun simplify(): EqnAstNode? {
        simplifyChildren()
        return if (operand()!!.type == Type.Constant) {
            EqnAstNodeDouble(evaluate())
        } else this
    }

    companion object {
        @kotlin.jvm.JvmStatic
        @Throws(EqnException::class)
        fun create(value: String, equationNode: EqnAstNode): EqnAstNodePreDefinedUnaryFunction {
            if (value == "sin") {
                return EqnAstNodeSin(equationNode)
            }
            if (value == "cos") {
                return EqnAstNodeCos(equationNode)
            }
            if (value == "tan") {
                return EqnAstNodeTan(equationNode)
            }
            if (value == "exp") {
                return EqnAstNodeExp(equationNode)
            }
            throw EqnException("PreDefinedUnaryFunction $value not yet implemented!")
        }
    }
}